server/player/player.go

```diff
@@ -2028,8 +2028,49 @@ func (p *Player) CollectExperience(value int) bool {
 	if time.Since(p.lastXPPickup.Load()) < time.Millisecond*100 {
 		return false
 	}
+
+	mendingItems := make([]item.Stack, 0, 6)
+	held, offHand := p.HeldItems()
+	if _, ok := offHand.Enchantment(enchantment.Mending{}); ok && offHand.Durability() < offHand.MaxDurability() {
+		mendingItems = append(mendingItems, offHand)
+	}
+	if _, ok := held.Enchantment(enchantment.Mending{}); ok && held.Durability() < held.MaxDurability() {
+		mendingItems = append(mendingItems, held)
+	}
+	for _, i := range p.Armour().Items() {
+		if i.Durability() == i.MaxDurability() {
+			continue
+		}
+		if _, ok := i.Enchantment(enchantment.Mending{}); ok {
+			mendingItems = append(mendingItems, i)
+		}
+	}
+	if length := len(mendingItems); length > 0 {
+		if foundItem := mendingItems[rand.Intn(length)]; foundItem.Durability() < foundItem.MaxDurability() {
+			repairAmount := math.Min(float64(foundItem.MaxDurability()-foundItem.Durability()), float64(value*2))
+			repairedItem := foundItem.WithDurability(foundItem.Durability() + int(repairAmount))
+			if repairAmount >= 2 {
+				// Mending removes 1 experience point for every 2 durability points. If the repaired durability is less than 2,
+				// then no experience is removed.
+				value -= int(math.Ceil(repairAmount / 2))
+			}
+			if offHand.Equal(foundItem) {
+				p.SetHeldItems(held, repairedItem)
+			} else if held.Equal(foundItem) {
+				p.SetHeldItems(repairedItem, offHand)
+			} else if slot, ok := p.Armour().Inventory().First(foundItem); ok {
+				_ = p.Armour().Inventory().SetItem(slot, repairedItem)
+			}
+		}
+	}
```

Comment:
I think it'd be a good idea to move this entire block into its own function. We'd want that to return the amount of leftover XP so we can use that for the rest of the code after it in this function.